### PR TITLE
API Implementation of /v2/attrs

### PR DESCRIPTION
**Pull request by daminichopra**
Fix for [Issue#118 ](https://github.com/smartsdk/ngsi-timeseries-api/issues/118) for Api implementation of /v2/attrs
Updated specification/quantumleap.yml for API method call.
Added query_NTNENA.py , query_NTNE1A.py for handling APIs.
Added test_NTNENA.py , test_NTNE1A.py for test cases related to API.

### CODE REVIEW

**Comment by c0c0n3** [comment]
@daminichopra great contribution, thank you so much!

I think we need more testing though to make sure we cover all the bases. Specifically, it'd be fantastic if you could implement these test scenarios:

1. Retrieve the same attribute `a` across two different entity types `T1`, `T2` for a specified time interval.
2. Given a time interval, retrieve two attributes `a`, `b` across two different entity types `T1`, `T2` where `T1.a` has the same type of `T2.a` and, likewise, `type (T1.b) = type (T2.b)`.
3. Average of a numeric attribute of `T1` and `T2` in a given time interval.
4. Average of a numeric attribute of `T1` and **non-numeric** attribute of another entity `T3`.

The database should initially contain entities for three different types `T1`, `T2`, `T3` and each type should have a string attribute `a`, `T1` and `T2` an integer attribute `b` whereas `T3` too should have an attribute called `b` but it should have a string type. Also `T1`, `T2`, `T3` should each have an attribute which is not in common with the other two types---e.g. `T1.x`, `T2.y`, `T3.z`. This way you should be able to use the same DB fixture for all four scenarios. In particular, you can use `T1.b` and `T2.b` for (3) whereas `T1.b` and `T3.b` should do for (4).

As usual these are just suggestions. Feel free to suggest variations on the theme or more scenarios or whatever you think would be appropriate. Oh also, sorry for the choice of entity/attr names, use whatever naming scheme/convention makes sense to you :-)

**Comment by daminichopra** [comment]
Hi  @c0c0n3 , I have Incorporated comments above in update PR.
Regarding points  3) and 4), we can consider them for future enhancement with new separate issue. As, we need to change code for this implementation and also it is not implemented for existing API yet. Also, we have an existing issue [Issue#241](https://github.com/smartsdk/ngsi-timeseries-api/issues/241) - Related to point above.

Please close this PR if it looks good . Thank you :)

**Comment by c0c0n3** [comment]
Hi @daminichopra thanks for implementing test scenarios (1) and (2). That's great 👍 
I agree with you about scenario (4), let's leave it behind for now. I opened a separate issue for it, #296. We still need to test (3) though: to me that's part of this PR, i.e. it should work without changing your code, if it doesn't...well perhaps we should fix it?

**Comment by chicco785** [comment]
from a logic point of view, as regards 4, simply there is no average. so you don't consider them.

**Comment by daminichopra** [comment]
Hi @c0c0n3 , I have Updated PR .Please check
Thank you !!!

**Comment by daminichopra** [comment]
@c0c0n3 @chicco785 , If PR is ok then Please close this PR.

**Comment by daminichopra** [comment]
Gentle Reminder!!

**Comment by c0c0n3** [comment]
Been busy busy w/ other stuff. Sorry. Will make some time for proper review tomorrow am, but casting a cursory glance... it looks ready to go :-)

**Comment by c0c0n3** [comment]
@daminichopra thanks for updating the PR with more tests, that's great.

When I looked at it yesterday I just assumed the failed tests where those for geocoding---we have an issue for that already, see #292. So you can ignore failures coming from `test_geocoding.py`.

But there are other tests that failed too, namely those in

* `src/tests/test_bc.py` 

Have a look at the [build log](https://travis-ci.org/github/smartsdk/ngsi-timeseries-api/builds/669184574#L1683) for the details. Did you get the green bar on your machine or did the tests fail on your box too? For good measure, always run

    $ source setup_dev_env.sh
    $ sh run_tests.sh

before committing. I know it takes one hell of a long time and we should actually improve that, but at the moment it's the only safety net we've got :-)

Can you please try reproducing on your box to see what's going on. Thanks!!

**Comment by daminichopra** [comment]
I have confirmed the same errors i.e. failing of `src/tests/test_bc.py`  in master branch in which this 
[PR](https://github.com/smartsdk/ngsi-timeseries-api/pull/294) is not included. Please check the below screen shot for your reference.
tests
tests1

**Comment by c0c0n3** [comment]
> have confirmed the same errors i.e. failing of src/tests/test_bc.py in master

Um, that's weird. PR #299 got a green bar but didn't touch that test or anything related. Perhaps we could try rebasing this PR onto master once I merge PR #299...

**Comment by daminichopra** [comment]
HI @c0c0n3 , Please merge this PR if it looks good. Thank you

File: specification/quantumleap.yml [hunk 1]
```diff
@@ -1349,8 +1349,8 @@ paths:
 
   /v2/attrs/{attrName}:
     get:
-      operationId: reporter.reporter.query_NTNE1A
-      summary: "(To Be Implemented) History of an attribute of N entities of N
+      operationId: reporter.query_NTNE1A.query_NTNE1A
+      summary: "History of an attribute of N entities of N
```
[the remark this answers is not in the document]
Comment (c0c0n3) [reply specification/quantumleap.yml:1353]: @daminichopra adding geo parameters to the Python method isn't enough, like I said in the comment above, `georel`, `geometry` and `coords` need to be declared in the `parameters` section of the `/v2/attrs/{attrName}` endpoint spec. Just can just copy and paste from e.g. `/v2/attrs`. Thanks alot!!!

**Comment by c0c0n3** [comment]
Hi @daminichopra, thank you so much for implementing geo params in Python :-)
Also, it looks like all tests are passing now, that's grand!

But there's still a couple of things that needs doing though before I can merge:

* Add geo params to the QL spec---see my comment in the review.
* Rebase this branch onto master.

Thanks!!

**Comment by daminichopra** [comment]
Hi @c0c0n3 , updated fix and rebase done. Please close this merge PR. Thank you

**Comment by c0c0n3** [comment]
@daminichopra thank you for the great contribution, we really appreciate your hard work and dedication!

**Pull request merged**